Feature/resonant axis calculator

idovgalyov-4xxi opened the pull request:
Description of the current version of the resonance calculator.

This version works with 2-body resonances for now.
The program contains several modules, main source file and makefile.

The modules are:
global_parameters - containing constants, simple functions and common parameters
id_matrix - module for managing id. matrices (creating, saving in memory, cleaning, etc.)
resonant_axis - calculates resonant values of semimajor axes needed for creating id. matrices (in files)
resonance_finder - finds possible resonances for an asteroid

The main source file "main.f95" is currently using for getting results by "resonance_finder" module.
Makefile is used for compiling and getting executable file "axis".
To run the program use the command `./axis asteroid a delta`, where:
- "asteroid" is an asteroid name
- "a"  - it's semimajor axis (in AU)
- "delta"  - semisize of searching interval (in AU)

The program finds possible resonances and places them in "asteroid.rpin" file.
It also creates files with id. matrices and checks possibility or their use.

Some comments:
- maximum resonance order currently gives as a parameter in "global_parameters" module.
- tests for .fun files are still in progress
- "main.f95" can be changed for testing some procedures if needed.

UPD:
- provided unit tests
- Did some changes in modules corresponding to review notes.

idovgalyov-4xxi committed:
Two-body case addition starting

Changed format to .f95 instead of .f90
In resonant_axis.f95 added function count_axis_2body
Provided test case in .fun-fule
The work is in progress.

idovgalyov-4xxi committed:
small update

Including makefile to be more compact

idovgalyov-4xxi committed:
Global parameters module

Added new module where we can contain global parameters, e.g. constants and planet physical parameters.
Main.f95 was modified to be able to create id-matrices for a given planet name (in capitals).
Example of running> ./axis NEPTUNE

Makefile was modified so as it creates correct executable file.

idovgalyov-4xxi committed:
Makefile update

Makefile was modified to be smarter with modules
It must be easy to add new modules now (requires recordinq new module name once only).
Also some changes in resonant_axis and global_parameters.

idovgalyov-4xxi committed:
resonance finder

Provided new modules:
id_matrix for managing id. matrices
resonance_finder for finding avaliable resonances for a current asteroid

main.f95 is needed for any launches and tests before we do a global interface

smirik commented on axis/global_parameters.f95:
```diff
@@ -0,0 +1,147 @@
+contains
+
+!----------------------------------------------------------------------------------------------
+integer function planet_id(s)
```
This function requires a unit test.

smirik commented on axis/global_parameters.f95:
```diff
@@ -0,0 +1,147 @@
+end function planet_id
+
+!----------------------------------------------------------------------------------------------
+function planet_name(i) result(s)
```
This function requires a unit test.

smirik commented on axis/global_parameters.f95:
```diff
@@ -0,0 +1,147 @@
+         case(9)
+             s='PLUTO'
+         case default
+             write(*,*)'Warning! Detected incorrect planet id, be careful.'
```
Better to say:
"Error! The planet id must be between 1 and 9."

idovgalyov-4xxi replied:
Done

smirik commented on axis/global_parameters.f95:
```diff
@@ -0,0 +1,147 @@
+            i=9
+        case default
+            i=0
+            write(*,*)'Warning! Detected incorrect planet name, be careful.'
```
Better to say:
"Error! Planet name must be one from the list: MERCURY, VENUS, EARTHMOO, MARS, JUPITER, SATURN, URANUS, NEPTUNE, PLUTO".

idovgalyov-4xxi replied:
Done

smirik commented on axis/global_parameters.f95:
```diff
@@ -0,0 +1,147 @@
+ end function planet_name
+
+!----------------------------------------------------------------------------------------------
+integer function nod(a,b)
```
This function requires a unit test.
Also, don't you think that the function requires too much variables? 
Here are the examples of more easy implementations:
https://users.soe.ucsc.edu/~pohl/FORTRAN/Chfor/gcd.for

smirik replied:
Also it should be named gcd, not nod.

idovgalyov-4xxi replied:
Refreshed the function, named gcd, deleted 1 variable and modified the algorithm.

smirik commented on axis/id_matrix.f95:
```diff
@@ -0,0 +1,110 @@
+contains
+
+!----------------------------------------------------------------------------------------------
+integer function get_idmatrix_2body_status(pl_id)
```
This function requires a unit test.

idovgalyov-4xxi replied:
There are still questions about unit test...

smirik commented on axis/id_matrix.f95:
```diff
@@ -0,0 +1,110 @@
+    integer:: s
+    integer:: pl_id,un
+
+    if(allocated(idmatrix_2body(pl_id)%matrix)) then
```
Could you please explain what happens here?

smirik commented on axis/id_matrix.f95:
```diff
@@ -0,0 +1,110 @@
+    integer:: pl_id,un
+
+    if(allocated(idmatrix_2body(pl_id)%matrix)) then
+        s=0
```
Isn't it a good idea to use return here instead of continuing the program execution?

idovgalyov-4xxi replied:
Modified this function for using "return" with no conflicts with Fortran specifications.

smirik commented on axis/id_matrix.f95:
```diff
@@ -0,0 +1,110 @@
+        open(unit=un,file='id_matrix_'//trim(pl_name)//'.dat',action='read',iostat=s)
+        if(s==0) then
+            close(un)
+            s=-1
```
See the comment above. Isn't return a good option?

smirik commented on axis/id_matrix.f95:
```diff
@@ -0,0 +1,110 @@
+! Produces:
+!   idmatrix_2body(pl_id)%matrix
+    integer:: pl_id,s,l,un
+    type(idmrow_2body),dimension(1000):: matrix
```
Why the dimension is 1000.

idovgalyov-4xxi replied:
Fixed this.
The method I've described.
Thought it would be acceptable.

smirik commented on axis/resonance_finder.f95:
```diff
@@ -0,0 +1,86 @@
+contains
+
+!----------------------------------------------------------------------------------------------
+subroutine get_all_possible_resonances_2body(asteroid,a,delta)
```
Why is it subroutine, not a function?

smirik replied:
OK

smirik commented on axis/resonance_finder.f95:
```diff
@@ -0,0 +1,86 @@
+                endif
+            enddo
+        case(-1)
+            write(*,*) 'Not found idmatrix in memory - try to find corresponding file...'
```
Please replace the text with the following:
"Cannot find the identification matrix in the memory.".

idovgalyov-4xxi replied:
Done

smirik commented on axis/resonance_finder.f95:
```diff
@@ -0,0 +1,86 @@
+        case(-1)
+            write(*,*) 'Not found idmatrix in memory - try to find corresponding file...'
+            un=8+pl_id
+            open(unit=un,file='id_matrix_'//trim(pl_name)//'.dat',action='read',iostat=s)
```
Why do you do here the loading manually, not via invocation of the function: add_idmatrix_2body.

idovgalyov-4xxi replied:
Justifiable.
Applied this subroutine.

smirik commented on axis/resonant_axis.f95:
```diff
@@ -0,0 +1,111 @@
+contains
+
+!----------------------------------------------------------------------------------------------
+subroutine make_id_matrix_2body(pl_name,max_order)
```
make -> build
This subroutine should have a unit test (that the file has been created & it is correct).

smirik replied:
Why this subroutine is in resonant_axis.f95, not in id_matrix?

idovgalyov-4xxi replied:
Renamed and moved to "id_matrix" module.
There are still questions about unit test...

smirik commented on axis/resonant_axis.f95:
```diff
@@ -0,0 +1,111 @@
+    open(unit=un,file="id_matrix_"//trim(pl_name)//".dat",status='replace')
+    ! Get id. matrix for a planet by a given maximum order
+    do m1=1,max_order
+        do m=-1,-m1-max_order,-1
```
Why m is negative only? 
5J+2 — IMO, normal resonance, is not it? In your case we check only 5J-2, right?

idovgalyov-4xxi replied:
Found some precedents.
Left as is for now.

smirik commented on axis/resonant_axis.fun:
```diff
@@ -36,6 +36,17 @@ test n_from_a_reverse_test
+    a1=5.20248019d0
+    m1=9.547919384243222d-4
+    a=count_axis_2body(resonance,a1,m1)
+    assert_equal_within(,2.8253d0,1d-3)
```
Add, also, please, well-known case for Kirkwood gaps.

idovgalyov-4xxi replied:
Done

smirik reviewed (changes requested):
Added comments. Most of them are related to the unit tests, but there are also few comments related to the functions / sub-routines.

smirik committed:
Add prettification to the code. Replace .f90 with .f95 in Makefile.

idovgalyov-4xxi committed:
Big update

Provided unit tests for the most part of functions and subroutines
Modified some functions according to review notes.
Nevertheless some specific aspects still require more conversations...

smirik reviewed (changes requested):
The tests are still not working, even if replace .f95 with .f90. The issue is in REAL(4) vs REAL(8). Need to be fixed before the commit.

smirik commented on axis/global_parameters.fun:
```diff
@@ -0,0 +1,54 @@
+    integer :: a1=5,b1=2,g1=1
+    integer :: a2=60,b2=12,g2=12
+    integer :: g
+    g=gcd(a1,b1)
```
IMO, additional things.
Why not to check directly?
`assert_equal(12,gcd(60,12))`?
Why do you create an extra variables?

smirik replied:
IMO, additional things.
Why not to check directly?
`assert_equal(12,gcd(60,12))`?
Why do you create an extra variables?

smirik commented on axis/id_matrix.f95:
```diff
@@ -0,0 +1,171 @@
+            read (un, *) idmatrix_2body(pl_id)%matrix(i)
+        enddo
+        close (un)
+! TO DO - here is a weakness of the program with uncertainity of matrix length - need a cure
```
Why the code is commented?

smirik commented on axis/id_matrix.fun:
```diff
@@ -0,0 +1,37 @@
+
+end test_suite
+
+! assert_true(expression)
```
These strings with the comments were added initially as examples. Please remove them!

idovgalyov-4xxi replied:
Done

smirik commented on axis/global_parameters.fun:
```diff
@@ -0,0 +1,54 @@
+
+end test_suite
+
+! assert_true(expression)
```
These strings with the comments were added initially as examples. Please remove them!

smirik commented on axis/id_matrix.f95:
```diff
@@ -0,0 +1,171 @@
+            read (un, *) idmatrix_2body(pl_id)%matrix(i)
+        enddo
+        close (un)
+! TO DO - here is a weakness of the program with uncertainity of matrix length - need a cure
```
Why the code is commented?

idovgalyov-4xxi replied:
Here I decided to use more stable file-measuring method.
I came to decision that it wouldn't be a significant fall of performance.

smirik commented on axis/main.f95:
```diff
@@ -0,0 +1,43 @@
+program resonant_axis_calculator
```
Just FYI: if I run the command without the arguments, it shows the error. But how can I find the arguments I should put?

idovgalyov-4xxi replied:
Provided Readme with description of how to lsunch the current version.

smirik commented on axis/resonance_finder.f95:
```diff
@@ -0,0 +1,93 @@
+                endif
+            enddo
+        case default
+            write (*, *) 'No founded idmatrix for ', pl_name, '!'
```
'No founded idmatrix' -> 'idmatrix was not found'

idovgalyov-4xxi replied:
Done

smirik reviewed (changes requested):
See the comments

idovgalyov-4xxi committed:
Big update #2

Corrected unit tests and returned to .f90 extension.
Some small corrections in source files
Created Readme.md (here will be descriptions and other sth)

idovgalyov-4xxi committed:
Update Makefile
